samza-core/src/main/scala/org/apache/samza/system/SystemConsumers.scala

```diff
@@ -192,10 +194,22 @@ class SystemConsumers (
 
     metrics.registerSystemStreamPartition(systemStreamPartition)
     unprocessedMessagesBySSP.put(systemStreamPartition, new ArrayDeque[IncomingMessageEnvelope]())
+
+    // Note regarding Startpoints and MessageChooser:
+    // Even if there is a startpoint for this SSP, passing in the checkpoint offset should not have any side-effects.
+    // Basically, the offset in the chooser is used in the special scenario where an SSP is both a broadcast and bootstrap stream
+    // and needs to decide what's the lowest starting offset for an SSP that spans across multiple tasks so it knows
+    // to keep the highest priority on the SSP starting from the lowest starting offset until the SSP is fully
+    // bootstrapped to the UPCOMING offset. The offset here is ignored otherwise.
     chooser.register(systemStreamPartition, offset)
 
     try {
-      consumers(systemStreamPartition.getSystem).register(systemStreamPartition, offset)
+      val consumer = consumers(systemStreamPartition.getSystem)
+      if (startpoint != null && consumer.isInstanceOf[StartpointVisitor]) {
+        startpoint.apply(systemStreamPartition, consumer.asInstanceOf[StartpointVisitor])
```

Comment:
I don't remember if this was discussed in the previous PR, but what do you think about adding a `SystemConsumer.register(systemStreamPartition, startpoint)` method? Then each implementation could provide their own visitor and call `startpoint.apply`.
The `StartpointVisitor` interface itself does not necessarily mean that something has to be registered. It just is a way to specify that something should happen to a `Startpoint`, so it is a little bit odd to use `StartpointVisitor` as a marker to denote that something should be registered.